feat[character] : 캐릭터 생성,조회 기능 추가

minji0828 opened the pull request:
## 📝 PR 요약
feat[character] : 캐릭터 생성,조회 기능 추가
패키지 구조 리팩터링


## 🔍 작업 상세 내용
헥사고날 느낌으로 패키지 구조 변경
캐릭터 생성 기능 추가 (최초 생성 시 기본 스탯 0으로 부여)
스탯은 VO로 존재

## ✅ 체크리스트
- [x] 의미 있는 커밋 컨벤션을 준수하며 작성했는가?
- [x] 변경된 코드에 대해 필요한 테스트를 추가했는가?
- [x] 로컬에서 정상적으로 빌드/실행되는지 확인했는가?
- [x] 불필요한 console.log, debug 코드를 제거했는가?
- [x] 주석은 필요한 곳만 남기고 정리했는가?

---

## 📸 화면(또는 기능) 변경 사항
<!-- UI 변경이 있으면 캡처 첨부해주세요. 없다면 '없음' -->
없음

---

## 🔗 관련 이슈
Closes #11

minji0828 committed:
feat[character] : 캐릭터 생성,조회 기능 추가

feat[character] : 캐릭터 생성,조회 기능 추가
패키지 구조 리팩터링

JSeHoone commented on build.gradle:
```diff
@@ -33,9 +33,7 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
     runtimeOnly 'com.mysql:mysql-connector-j'
     annotationProcessor 'org.projectlombok:lombok'
-    testImplementation 'org.springframework.boot:spring-boot-starter-data-jpa-test'
-    testImplementation 'org.springframework.boot:spring-boot-starter-mail-test'
-    testImplementation 'org.springframework.boot:spring-boot-starter-webmvc-test'
```
Test 모듈 쪽 의존성이 사라진 것 같습니다 !

minji0828 committed:
Update build.gradle

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Character.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Id
+    @Column(columnDefinition = "BINARY(16)")
+    private UUID id;
```
캐릭터 ID 를 Long 타입의 자동 생성 타입이 아닌, UUID 로 사용하신 이유가 있으신가요?

전반적인 성능 관점에서 Long 타입 auto increment 키가 더 안정적이고, 성능 측면에서도 더 유리하다고 생각합니다.
어떻게 생각하시나요?

SeungYeop-Han replied:
Long 으로 변경 하신 것 확인 했습니다. UUID 사용이 잘못되었다기 보다는 사용 이유가 궁금했던 것 뿐이었습니다. 나중에 이야기 나눠봐요!

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/application/service/CharacterService.java:
Service도 interface로 만들어서 구현체랑 구분한 점이 너무 좋습니다 !

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/presentation/controller/CharacterController.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/characters")
```
여기서 `ReqeustMapping("/api/characters")`로 바꿔주시면 좋을 것 같습니다 !

minji0828 replied:
이 pr에서 수정하겠습니다.

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/application/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public Character get(UUID id) {
+        return repository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("Character not found: " + id));
```
`IllegalArgumentException` 은 전달하는 아규먼트가 유효하지 않은 경우에 발생시키는 편이 자연스럽습니다.

위 경우는 해당 id 형식 자체는 유효하지만, 단지 존재하지 않을 뿐이므로 일단 `RuntimeException` 사용하시거나 XxxNotFoundException 과 같은 도메인 예외 별도 정의 후 사용하시는 건 어떻게 생각하시나요?

sobee1403 replied:
XxxNotFoundException에 한표 던집니다

SeungYeop-Han replied:
RuntimeException 으로 변경하신 것 확인했습니다! 당장은 아니지만 나중에 팀 회의에서 논의되었던 대로 도메인 예외로 바꿔봅시다!

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/application/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+import com.potenhoon.tododoapi.character.domain.model.Stats;
+import com.potenhoon.tododoapi.character.domain.repository.CharacterRepository;
+import com.potenhoon.tododoapi.character.application.service.CharacterService;
+import org.springframework.stereotype.Service;
+
+import java.util.UUID;
+
+@Service
+public class CharacterServiceImpl implements CharacterService {
+
+    private final CharacterRepository repository;
+
+    public CharacterServiceImpl(CharacterRepository repository) {
+        this.repository = repository;
+    }
+
+    @Override
+    public Character create(String name, Stats stats) {
+        Character character = Character.create(name, stats);
+        return repository.save(character);
+    }
+
+    @Override
+    public Character get(UUID id) {
+        return repository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("Character not found: " + id));
+    }
+}
```
이 부분에서는 추후 정책이나 검증로직에 대해서 주석을 달아주시면 좋을 것 같습니다 !

그리고 throw발생 시키는 부분에서도 TODO를 남겨서 추후 GlobalException을 적용해야한다는 주석을 명시적으로 남겨주시면 좋을 것 같습니다 !

minji0828 replied:
수정하겠습니다.

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Character.java:
```diff
@@ -0,0 +1,59 @@
+    private String name;
+
+    @Embedded
+    private Stats stats;
```
`@Embedded` 어노테이션을 사용하면 테이블의 컬럼이 생기는 것으로 알고 있습니다 !

stat이 많아지게 되면 Character 테이블의 컬럼도 증가하게 되어서 조회 성능이 떨어질 수도 있을 것 같습니다

Stat테이블을 이용해서 oneToOne과 같은 관계 설정을 주는 것도 좋아보입니다 ! 이 부분은 추후에 팀 회의를 통해서 확립하는 것이 좋아보입니다 !

SeungYeop-Han replied:
어차피 캐릭터 하나에 스탯 한 세트만 있으니까 이 부분은 괜찮지 않을까요?

minji0828 replied:
음 일단 메모해두고 추후에 얘기하도록 하죠

sobee1403 replied:
저도 추후에 스탯의 종류가 추가될 것을 고려하여 관계설정을 해주는 편이 좋다고 생각합니도

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Character.java:
```diff
@@ -0,0 +1,59 @@
+        Stats resolved = stats == null ? Stats.zero() : stats;
+        return new Character(UUID.randomUUID(), name, resolved);
+    }
+
```
오버로딩을 사용해서 다형성을 만든 것이 짱입니다 !

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/infra/jpa/JPACharacterRepository.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.UUID;
+
+@Repository
+public class JPACharacterRepository implements CharacterRepository {
```
- JPA -> Jpa (컨벤션)
- `CharacterJpaRepository` 는 인터페이스 이름이고 `JPACharaterRepository` 는 `CharactorRepository` 의 구현체인데, 이름만 보면 어떤게 어떤 역할을 하는건지 잘 와닿지 않습니다. `JPACharacterRepository` 의 이름을 예를 들면 `CharacterRepositoryJpaAdapter` 와 같은 식으로 바꾸는 건 어떻게 생각하시나요?

SeungYeop-Han replied:
확인했습니다!

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/application/service/CharacterService.java:
```diff
@@ -0,0 +1,13 @@
+
+    Character create(String name, Stats stats);
+
+    Character get(UUID id);
```
저번 회의 때 BC 외부로 엔티티 노출을 피하자는 논의가 있었습니다.
Character 엔티티 객체를 반환하는 것이 아니라 Dto 를 반환하는 건 어떤가요?

SeungYeop-Han replied:
변경 하신 부분 확인했습니다!

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/infrastructure/jpa/CharacterJpaRepository.java:
`CharacterJpaRepository`랑 `JPACharacterRepository`랑 중복되는 것 같습니다 !

```java
public interface CharacterJpaRepository extends JpaRepository<Character, UUID>, CharacterRepository {
}
```
이렇게 domain/repository에서 정의한 repository랑 JPA를 동시에 받아올 수 있습니다 !

minji0828 replied:
수정하겠습니다

SeungYeop-Han replied:
제 생각에는 민수님이 하신 것 처럼 어댑터 방식(member 부분 수업 때 사용한 방식) 도 괜찮아 보입니다!

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/presentation/controller/CharacterController.java:
```diff
@@ -0,0 +1,45 @@
+
+    private Stats resolveStats(StatsRequest statsRequest) {
+        return statsRequest == null ? Stats.zero() : statsRequest.toStats();
+    }
```
해당 메서드는 CharacterCreateRequest 내부에서 검증해주는 것이 좋아보입니다 !

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/presentation/dto/CharacterResponse.java:
```diff
@@ -0,0 +1,12 @@
+public record CharacterResponse(UUID id, String name, Stats stats) {
+    public static CharacterResponse from(Character character) {
+        return new CharacterResponse(character.getId(), character.getName(), character.getStats());
+    }
```
presentation 레이어에 도메인 엔티티(`Character`) 노출됨

JSeHoone replied:
승엽님이 말씀해주신 내용이 반영되지 않은 것 같습니다 !

JSeHoone reviewed (commented):
infra 패키지로 만들어주셨는데, intrastructure로 패키지 명을 변경해주시면 좋을 것 같습니다 !

SeungYeop-Han commented on src/main/java/com/potenhoon/tododoapi/character/presentation/dto/StatsRequest.java:
```diff
@@ -0,0 +1,9 @@
+import com.potenhoon.tododoapi.character.domain.model.Stats;
+
+public record StatsRequest(int a, int b, int c, int d, int e, int f) {
+    public Stats toStats() {
+        return new Stats(a, b, c, d, e, f);
+    }
```
presentation 레이어에 도메인 엔티티(`Stats`) 노출됨

SeungYeop-Han commented on build.gradle:
```diff
@@ -33,9 +33,7 @@ dependencies {
-    testImplementation 'org.springframework.boot:spring-boot-starter-data-jpa-test'
-    testImplementation 'org.springframework.boot:spring-boot-starter-mail-test'
-    testImplementation 'org.springframework.boot:spring-boot-starter-webmvc-test'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
build.gradle 수정하신 이유가 있으신가요?

minji0828 replied:
잘못수정한것입니다

SeungYeop-Han replied:
확인했습니다!

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Character.java:
```diff
@@ -0,0 +1,59 @@
+
+    private Character(UUID id, String name, Stats stats) {
+        this.id = Objects.requireNonNull(id, "id must not be null");
+        this.name = Objects.requireNonNull(name, "name must not be null").trim();
```
trim처리는 캐릭터 네임의 정책일까요?

minji0828 replied:
그냥 공백제거용입니다

JSeHoone replied:
캐릭터 네임을 trim 처리하는 것도 정책이 될 수 있습니다.
`새로운 이름` 이라는 이름을 가진 캐릭터를 생성할 수 도 있는데, trim 처리를 해버리면 `새로운이름`으로 생성 되어서 사용자의 의도대로 되지 않을 수 있습니다. 

 이 부분은 정책적으로 논의 후에 trim()을 추가할지 삭제할지 결정하면 좋을 것 같습니다.

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Stats.java:
```diff
@@ -0,0 +1,62 @@
+@Embeddable
+public class Stats {
+
+    @Column(name = "stat_a")
```
정확한 명칭 부여가 아닌 임의의 a, b, c같은 걸로 명칭 부여한 이유가 있을까요?

SeungYeop-Han replied:
아마도 스탯을 어떤 걸로 할 지 정해지지 않아서 그러신 것 같습니다.

sobee1403 replied:
스탯을 어떤걸로 할지 정해지지 않아도 해당 부분은 리스트로 받거나, 
아니면 현재는 단일 객체로 생성하여 관리할 수 있는 부분이라는 생각이 듭니다.
포괄적이고 불분명한 명칭은 문제가 될 수 있다고 생각됩니다.

minji0828 replied:
이거는 일단 단일객체로 만들어놓을게요

minji0828 committed:
fix[character] 외부 노출된 domain 수정

pr 피드백 반영 수정
-ReqeustMapping("/api/characters") 수정
JPA infra수정
resolvestats 제거
캐릭터 id  uuid=>Long타입 수정

JinYong833 commented on src/main/java/com/potenhoon/tododoapi/character/application/service/CharacterService.java:
```diff
@@ -1,13 +1,13 @@
+    CharacterResponse create(CharacterCreateRequest request);
 
-    Character get(UUID id);
+    CharacterResponse get(UUID id);
```
매개변수가 UUID 입니다. 확인부탁드려요

minji0828 replied:
반영함

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Stats.java:
```diff
@@ -0,0 +1,66 @@
+        this.f = f;
+    }
+
+    public static Stats getStats(Stats stats) {
```
지금 구현하신 정적 팩토리 메서드 getStats(Stats stats)는
결론적으로 입력으로 받은 Stats 객체의 필드를 그대로 복사해서 새로운 Stats 객체를 리턴하는 형태입니다.
하지만 현재 형태로는 다음과 같은 문제가 있습니다.

1. 의미적 이득이 없습니다.
외부에서 넘겨준 Stats와 동일한 값을 가진 새 객체를 만드는 것만으로는
별도의 생성 규칙, 유효성 체크, 변환 로직 등 ‘정적 팩토리 메서드가 있어야 할 이유’가 보이지 않습니다.

2. 불필요한 객체 생성으로 자원 낭비가 발생합니다.
내부 상태 변경이나 방어적 복사목적이 아니라면,
동일한 객체를 새로 생성하는 작업은 GC pressure만 증가시키게 됩니다.

3. 사용 목적이 모호합니다.
지금 형태라면 getStats(stats) 대신 그냥 stats를 그대로 넘겨도 로직적으로 차이가 없습니다.
즉, 이 메서드를 사용해야 하는 명확한 상황이 보이지 않습니다.

혹시라도  “복사본 생성”이 필요하다면 메서드명과 목적을 명확히 해주시기 바랍니다.
ex) copyOf(Stats source)

만약 방어적 복사, 불변성 확보, 일부 값 변환 등의 이유가 아니라면
메서드 자체를 제거하는 것이 더 명확하고 효율적일 수 있습니다.

minji0828 replied:
반영함

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/application/service/CharacterService.java:
```diff
@@ -0,0 +1,13 @@
+
+    CharacterResponse create(CharacterCreateRequest request);
+
+    CharacterResponse get(UUID id);
```
서비스 명칭이 모호한 것 같습니다 조금더 직관적인 네이밍이 필요합니다!

minji0828 replied:
확인

minji0828 replied:
반영함

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/application/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public CharacterResponse get(UUID id) {
+        var found = repository.findById(id)
```
var은 타입을 숨겨서 가독성을 떨어뜨리고, 협업자가 타입을 추적해야 하므로 협업시 문제가 될 수 있는 부분입니다.
findById는 반환 타입을 직관적으로 알기 어려우므로 명시적 타입 선언이 더 안전하다고 봅니다...!

minji0828 replied:
확인

minji0828 replied:
반영함

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/application/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public CharacterResponse get(UUID id) {
+        var found = repository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Character not found: " + id));
```
상황에 맞는 custom exception을 추후에 도입해주실거라고 믿습니다...!

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/presentation/dto/CharacterResponse.java:
record+from 패턴으로 잘 구현되어있고 깔끔한 DTO였습니다
굿굿

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/presentation/dto/StatsRequest.java:
toStats와 toStatsOrZero 메서드의 차이가 있을까요?
결과적으로 같은 기능을 하는 메서드로 보입니다

만약 toStatesOrZero 명칭을 보아 Stats 객체를 리턴하거나, 혹은 적절하지 않은 값이 있을 경우
zero인 Stats 객체를 리턴해줘야하는 것으로 보이는데, 그럴경우 조건문으로 유효성을 검증하여 유효하지 않으면
return Stats.zero(); 같은 식으로 리턴 값을 분기해줘야하지 않을까요?

minji0828 replied:
반영 예정

minji0828 replied:
반영함

minji0828 committed:
fix[character] stats entity 분리

stats 별도 테이블로 캐릭터와 1대1 관계로 분리

캐릭터 ID 타입 UUID -> Long 타입으로 변경

모호한 서비스명칭 수정

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Stats.java:
스탯 하나만 남겨두기

minji0828 replied:
반영함

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/application/service/CharacterService.java:
```diff
@@ -1,13 +1,11 @@
+    CharacterResponse createCharacter(CharacterCreateRequest request);
 
-    CharacterResponse get(UUID id);
+    CharacterResponse getCharacterById(Long id);
```
이 부분은 주석으로 남겨두기

minji0828 replied:
반영함

minji0828 committed:
fix[character] Stats column 수정

 Stats column 하나로 수정
컬럼 명 stat 으로 수정

getCharacterById Long id 필요도 없고 의미도 없어
UUID userId로 조회하는 것으로 변경

minji0828 committed:
fix[Character] stats dto 수정

a, b, c -> stat로 수정

minji0828 committed:
fix[Character]stats return 수정

0 -> stat 수정

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Stats.java:
```diff
@@ -1,17 +1,19 @@
     @Id
-    @Column
     @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
```
테이블 관점에서 보면 `id`라는 컬럼명보다는 `stat_id`라는 컬럼명이 좋아보이는데, 어떻게 생각하시나요 ?

minji0828 replied:
반영함

JSeHoone commented on src/main/java/com/potenhoon/tododoapi/character/presentation/controller/CharacterController.java:
```diff
@@ -0,0 +1,38 @@
+        return ResponseEntity.ok(created);
+    }
+
+    @GetMapping("/{id}")
+public ResponseEntity<CharacterResponse> getCharacterById(@PathVariable UUID userId) {
+    CharacterResponse found = characterService.getCharacterById(userId);
+        return ResponseEntity.ok(found);
+    }
```
캐릭터 세부 조회 API를 구현하신 것 같은데, PathVariable로 user_id를 받아오면 URL에 user_id가 노출되어서 보안적으로 취약해질 것 같습니다. 

테스트를 위해서 넣어주신 거면 주석을 남겨주시면 좋을 것 같습니다 !

minji0828 replied:
반영함

minji0828 committed:
fix[character] stats 컬럼명 변경

id - > stat_id

그 외 특이사항 주석처리

JinYong833 commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Stats.java:
```diff
@@ -0,0 +1,41 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Long id;
+
```
캐릭터 ID 필요하지 않을까요?

minji0828 committed:
fix[character] stat id 제거

stat이 캐릭터 id를 갖게 변경

minji0828 committed:
fix[character] 캐릭터 도메인 엔티티 외부노출 x

캐릭터 도메인 엔티티 외부노출 안하게 정적 팩토리 메서드

도메인 - dto 전환용 매핑 추가

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/application/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public CharacterResponse createCharacter(CharacterCreateRequest request) {
+        Stats stats = request.stats() == null ? Stats.zero() : request.stats().toStats();
```
현재 서비스 레이어에서 request.stats()가 null인지 확인한 뒤 Stats.zero()를 적용하고 있는데요,
이런 기본값 설정 로직을 도메인 객체 내부로 옮기면 더 응집도 높은 구조가 될 것 같습니다.

Lombok의 @Builder.Default 등을 활용해 Stats 내에서 기본값을 정의해두면,
서비스 레이어에서는 별도의 null 체크 없이 도메인 생성만 호출하면 되기 때문에
“기본값 → Stats가 책임지는 로직”으로 캡슐화할 수 있을 것 같아요.

sobee1403 commented on src/main/java/com/potenhoon/tododoapi/character/domain/model/Character.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public static Character create(String name, UUID userId, Stats stats) {
+        Stats resolved = stats == null ? Stats.zero() : stats;
```
이 로직도 Stats.stat의 디폴트값을 0으로 세팅해준다면 필요하지 않은 로직이라 삭제가 가능할것 같아요

minji0828 committed:
fix[character] stats 안에서 기본으로 0을 부여

외부에서 기본스탯을 0으로 주는게 아니고
stats 자체에서 기본으로 0을 주게 @builder.default 로 적용
